### ISSUE TITLE
Builds are randomly failing ("Bad Request" error in worker logs)

### ISSUE THREAD

**Issue by aplex** [Microsoft/vsts-agent#424]
## **Agent version and platform**
Version of your agent? 2.103.0
<br>
OS X
## **VSTS type and version**
On-Prem TFS 2015 Update 2
<br>

## **What's not working?** 
After migrating to vsts agent (from old vso agent), builds are randomly failing on start from time to time. In TFS, build is shown as failed, no build steps were executed. TFS shows that build ran for 5 seconds.

This affects different build definitions. 
Source control is TFVC. 
This machine has two build agents running in parallel from different folders.
<br>
## **Agent and Worker's diag log**

Worker's log:

```

[2016-07-20 21:56:56Z INFO ConfigurationStore] currentAssemblyLocation: /Users/admin/build_agents/mac_agent_3/bin/Agent.Worker.dll
[2016-07-20 21:56:56Z INFO ConfigurationStore] binPath: /Users/admin/build_agents/mac_agent_3/bin
[2016-07-20 21:56:56Z INFO ConfigurationStore] RootFolder: /Users/admin/build_agents/mac_agent_3
[2016-07-20 21:56:56Z INFO ConfigurationStore] ConfigFilePath: /Users/admin/build_agents/mac_agent_3/.agent
[2016-07-20 21:56:56Z INFO ConfigurationStore] CredFilePath: /Users/admin/build_agents/mac_agent_3/.credentials
[2016-07-20 21:56:56Z INFO ConfigurationStore] ServiceConfigFilePath: /Users/admin/build_agents/mac_agent_3/.service
[2016-07-20 21:56:56Z INFO JobRunner] Ensure endpoint url match config url base. https://devtfs.company.net/tfs/DefaultCollection/
[2016-07-20 21:56:56Z INFO JobRunner] Ensure System.TaskDefinitionsUrl match config url base. https://devtfs.company.net/tfs/DefaultCollection/
[2016-07-20 21:56:56Z INFO JobRunner] Ensure System.TFCollectionUrl match config url base. https://devtfs.company.net/tfs/DefaultCollection/
[2016-07-20 21:56:56Z INFO JobRunner] Ensure SystemConnection url match config url base. https://devtfs.company.net/tfs/DefaultCollection/
[2016-07-20 21:56:56Z INFO JobRunner] Creating job server with URL: https://devtfs.company.net/tfs/DefaultCollection/
[2016-07-20 21:56:56Z INFO Worker] Listening for cancel message from the channel.
[2016-07-20 21:56:56Z INFO Worker] Waiting for the job to complete or for a cancel message from the channel.
[2016-07-20 21:56:57Z INFO Worker] Job completed.
[2016-07-20 21:56:57Z ERR  Program] Microsoft.VisualStudio.Services.WebApi.VssServiceResponseException: Bad Request
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.HandleResponse(HttpResponseMessage response)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__45.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.<SendAsync>d__42`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.VisualStudio.Services.Location.Client.LocationHttpClient.<GetConnectionDataAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.VisualStudio.Services.WebApi.Location.VssServerDataProvider.<ConnectAsync>d__41.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.JobServer.<ConnectAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.Worker.JobRunner.<RunAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.Worker.Worker.<RunAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.Worker.Program.<MainAsync>d__1.MoveNext()
```
Agent log:

```

[2016-07-20 21:56:50Z INFO JobDispatcher] Job request e9a8a101-72b1-450d-9c13-e49d405315d7 received.
[2016-07-20 21:56:50Z INFO JobDispatcher] Job request b455f84c-7fb4-4195-b272-b5119443dbaf processed succeed.
[2016-07-20 21:56:50Z INFO Terminal] WRITE LINE: 2016-07-20 21:56:50Z: Running job: Build
[2016-07-20 21:56:50Z INFO JobNotification] Entering JobStarted Notification
[2016-07-20 21:56:50Z INFO JobDispatcher] Start renew job request.
[2016-07-20 21:56:50Z INFO JobDispatcher] Successfully renew job request, job is valid till 7/20/16 10:01:50 PM
[2016-07-20 21:56:50Z INFO ProcessInvoker] Starting process with file name '/Users/admin/build_agents/mac_agent_3/bin/Agent.Worker', arguments 'spawnclient 87 90', and working directory '/Users/admin/build_agents/mac_agent_3/bin'.
[2016-07-20 21:56:50Z INFO ProcessInvoker] Process started with process id 19989, waiting for process exit.
[2016-07-20 21:56:50Z INFO JobDispatcher] Send job request message to worker.
[2016-07-20 21:56:53Z INFO ProcessInvoker] Finished process with exit code 1, and elapsed time 00:00:02.8340987.
[2016-07-20 21:56:53Z INFO JobDispatcher] Worker finished. Code: 1
[2016-07-20 21:56:53Z INFO JobDispatcher] finish job request with result: Failed
[2016-07-20 21:56:53Z INFO Terminal] WRITE LINE: 2016-07-20 21:56:53Z: Job Build completed with result: Failed
[2016-07-20 21:56:53Z INFO JobNotification] Entering JobCompleted Notification
[2016-07-20 21:56:53Z INFO JobDispatcher] Stop renew job request.
[2016-07-20 21:56:53Z INFO JobDispatcher] job renew has been canceled, stop renew job request.
[2016-07-20 21:56:55Z INFO JobDispatcher] Job request 0ece137d-ed79-4b3d-b9b3-57d781f40518 received.
[2016-07-20 21:56:55Z INFO JobDispatcher] Job request e9a8a101-72b1-450d-9c13-e49d405315d7 processed succeed.
[2016-07-20 21:56:55Z INFO Terminal] WRITE LINE: 2016-07-20 21:56:55Z: Running job: Build
[2016-07-20 21:56:55Z INFO JobNotification] Entering JobStarted Notification
[2016-07-20 21:56:55Z INFO JobDispatcher] Start renew job request.
[2016-07-20 21:56:55Z INFO JobDispatcher] Successfully renew job request, job is valid till 7/20/16 10:01:55 PM
[2016-07-20 21:56:55Z INFO ProcessInvoker] Starting process with file name '/Users/admin/build_agents/mac_agent_3/bin/Agent.Worker', arguments 'spawnclient 87 90', and working directory '/Users/admin/build_agents/mac_agent_3/bin'.
[2016-07-20 21:56:55Z INFO ProcessInvoker] Process started with process id 19991, waiting for process exit.
[2016-07-20 21:56:55Z INFO JobDispatcher] Send job request message to worker.
[2016-07-20 21:56:57Z INFO ProcessInvoker] Finished process with exit code 1, and elapsed time 00:00:01.9609924.
[2016-07-20 21:56:57Z INFO JobDispatcher] Worker finished. Code: 1
[2016-07-20 21:56:57Z INFO JobDispatcher] finish job request with result: Failed
[2016-07-20 21:56:57Z INFO Terminal] WRITE LINE: 2016-07-20 21:56:57Z: Job Build completed with result: Failed
[2016-07-20 21:56:57Z INFO JobNotification] Entering JobCompleted Notification
[2016-07-20 21:56:57Z INFO JobDispatcher] Stop renew job request.
[2016-07-20 21:56:57Z INFO JobDispatcher] job renew has been canceled, stop renew job request.
```

**Comment by TingluoHuang** [comment]
@aplex Can you set up Charles proxy to catch http trace while this happen? https://github.com/Microsoft/vsts-agent/blob/master/docs/troubleshooting.md#http-tracing-osx--linux

**Comment by aplex** [comment]
Ok, I was able to trap the interaction between tfs and vsts agent on OS X.

Multiple requests worked fine until I started Windows and OS X builds in parallel (don't know if this is a coincidence or they interfere with each other).

On successful build attempts, build agent makes a request that immediately succeeds:
```
GET https://devtfs.company.net/tfs/DefaultCollection/_apis/connectionData?connectOptions=1&lastChangeId=-1&lastChangeId64=-1
```

On a failed build attempt, the same request returned 401 error from server:

```
GET https://devtfs.company.net/tfs/DefaultCollection/_apis/connectionData?connectOptions=1&lastChangeId=-1&lastChangeId64=-1 HTTP/1.1
Host: devtfs.company.net
Accept: */*
Accept-Encoding: gzip
User-Agent: VSServices/15.255.65000.0 (NetStandard) VstsAgentCore-osx.10.11-x64/2.103.0 (Darwin 15.6.0 Darwin Kernel Version 15.6.0: Thu Jun 23 18:25:34 PDT 2016; root:xnu-3248.60.10~1/RELEASE_X86_64)
Accept-Language: en-US
X-TFS-FedAuthRedirect: Suppress
X-TFS-Session: ZZZ
Expect: 100-continue
Authorization: Bearer XXX


HTTP/1.1 401 Unauthorized
Content-Type: text/html
Server: Microsoft-IIS/8.5
X-TFS-ProcessId: 0214947e-5c31-4aaa-85d8-669f3ab8465e
Access-Control-Allow-Origin: *
Access-Control-Max-Age: 3600
Access-Control-Allow-Methods: OPTIONS,GET,POST,PATCH,PUT,DELETE
Access-Control-Expose-Headers: ActivityId,X-TFS-Session,X-MS-ContinuationToken
Access-Control-Allow-Headers: authorization
X-FRAME-OPTIONS: SAMEORIGIN
Set-Cookie: Tfs-SessionId=ZZZ; path=/; secure
Set-Cookie: Tfs-SessionActive=2016-07-21 21:22:56Z; path=/; secure
WWW-Authenticate: Basic realm="devtfs.company.net"
WWW-Authenticate: NTLM
P3P: CP="CAO DSP COR ADMa DEV CONo TELo CUR PSA PSD TAI IVDo OUR SAMi BUS DEM NAV STA UNI COM INT PHY ONL FIN PUR LOC CNT"
X-Content-Type-Options: nosniff
Date: Thu, 21 Jul 2016 21:22:56 GMT
Content-Length: 1293

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
<title>401 - Unauthorized: Access is denied due to invalid credentials.</title>
<style type="text/css">
<!--
body{margin:0;font-size:.7em;font-family:Verdana, Arial, Helvetica, sans-serif;background:#EEEEEE;}
fieldset{padding:0 15px 10px 15px;} 
h1{font-size:2.4em;margin:0;color:#FFF;}
h2{font-size:1.7em;margin:0;color:#CC0000;} 
h3{font-size:1.2em;margin:10px 0 0 0;color:#000000;} 
#header{width:96%;margin:0 0 0 0;padding:6px 2% 6px 2%;font-family:"trebuchet MS", Verdana, sans-serif;color:#FFF;
background-color:#555555;}
#content{margin:0 0 0 2%;position:relative;}
.content-container{background:#FFF;width:96%;margin-top:8px;padding:10px;position:relative;}
-->
</style>
</head>
<body>
<div id="header"><h1>Server Error</h1></div>
<div id="content">
 <div class="content-container"><fieldset>
  <h2>401 - Unauthorized: Access is denied due to invalid credentials.</h2>
  <h3>You do not have permission to view this directory or page using the credentials that you supplied.</h3>
 </fieldset></div>
</div>
</body>
</html>
```

After that, build agent sent two Authorization headers, and this resulted in 400 Bad request error:

```

GET https://devtfs.company.net/tfs/DefaultCollection/_apis/connectionData?connectOptions=1&lastChangeId=-1&lastChangeId64=-1 HTTP/1.1
Host: devtfs.company.net
Authorization: NTLM YYY
Accept: */*
Accept-Encoding: gzip
User-Agent: VSServices/15.255.65000.0 (NetStandard) VstsAgentCore-osx.10.11-x64/2.103.0 (Darwin 15.6.0 Darwin Kernel Version 15.6.0: Thu Jun 23 18:25:34 PDT 2016; root:xnu-3248.60.10~1/RELEASE_X86_64)
Accept-Language: en-US
X-TFS-FedAuthRedirect: Suppress
X-TFS-Session: ZZZ
Expect: 100-continue
Authorization: Bearer XXX


HTTP/1.1 400 Bad Request
Content-Type: text/html; charset=us-ascii
Server: Microsoft-HTTPAPI/2.0
Date: Thu, 21 Jul 2016 21:22:56 GMT
Connection: close
Content-Length: 339

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Bad Request</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Bad Request - Invalid Header</h2>
<hr><p>HTTP Error 400. The request has an invalid header name.</p>
</BODY></HTML>

```

**Comment by TingluoHuang** [comment]
#223 same issue.

**Comment by aplex** [comment]
This is a pretty bad bug, and we hit it every day. Is there any workaround or an ETA for a fix?

**Comment by TingluoHuang** [comment]
@aplex this might be a bug on dotnet core itself, Can you send me a full http trace from the time before the first 401 happen to the 400 happen? there should be some requests other than the two you paste here.

**Comment by aplex** [comment]
There were hundreds of requests there, and I don't have them now. I can try to re-capture them next time. 
Interesting thing is why did TFS respond with 401 in the first place. There was a token in that request.

**Comment by TingluoHuang** [comment]
@aplex that token is only valid for a short period time about an hour, server will response 401 to client so the client will request for a new token.

**Comment by TingluoHuang** [comment]
@aplex  we had a idea about what might cause the problem, but it just a guess, since I can't repro this, my PR #434  will contain the change i am talking about.

**Comment by TingluoHuang** [comment]
@aplex We published 2.104.1 which contains the change might able to fix your problem, can you take a try? Thanks,

**Comment by aplex** [comment]
I've installed new version, works fine so far.
